### PR TITLE
Implementation of a calibration pipeline using the wire grid

### PR DESCRIPTION
**Pull request by hnakata-JP**
I've just added functions that are used for the det_angle calibration with the wire grid.
Also, this commit has a tutorial just to check them temporarily. I will move this tutorial to the pwg-tutorials repo.

Anyway, I would be happy if someone could check these functions and give me comments to update them before merging them into the master branch.

Maybe do I need to wrap these functions with a class?

### CODE REVIEW

**Comment by hnakata-JP** [comment]
I added a configuration file for these functions. We also need a configuration/context file of detmatch and HWP as well.

**Comment by msilvafe** [comment]
> I added a configuration file for these functions. We also need a configuration/context file of detmatch and HWP as well.

For configuration files: The configuration files should not go in here. In the long run I think they'll end up in [site-pipeline-configs](https://github.com/simonsobs/site-pipeline-configs).

For context files: For metadata outside of or combinations of what's listed in the [Metadata Registry](https://simonsobs.atlassian.net/wiki/spaces/PRO/pages/222593410/Metadata+Registry) I have also been making my own context files and been keeping them in `/so/home/msilvafe/shared_files/contexts` I think that's a good solution or hosting them in [pwg-scripts](https://github.com/simonsobs/pwg-scripts). I think in the long run when we freeze a context that's widely useful we will host these in `/so/metadata/<platform>/contexts` (on site computing at least).

For the calibration script: This should go in [site_pipeline](https://github.com/simonsobs/sotodlib/tree/master/sotodlib/site_pipeline) and be called `analyze-wiregrid` at least according to the [FLP design document](https://docs.google.com/document/d/1Hb0LMDB7oUXugeHKqMLOq8EvVijHvPN7ZgB4yZgIEaU/edit#heading=h.hoe7hjy6kth). Take a look at the structure of some of the other site_pipeline elements to see how they're structured. In particular you want main to actually be defined with arguments passed by argparse (or similar) as we want this to be callable from the command line (probably automatically with prefect at some point). I'll put some other comments inline as well.

I'll add a few more inline comments but I'm happy to see this code committed and the wire grid system working so well! I'm looking forward to going through the tutorials soon!

File: sotodlib/wiregrid/calibrate.py
Comment (msilvafe) [inline sotodlib/wiregrid/calibrate.py:21]: These should not be defined here. If you need this info (although I don't see it used so I think it might be a leftover from your testing) get it from the axis manager or metadata.

File: sotodlib/wiregrid/calibrate.py
Comment (msilvafe) [inline sotodlib/wiregrid/calibrate.py:23]: Typically keyword arguments are passed in this way if you want to pass a set of optional arguments to a function called inside of this function. For the default parameters for this function you should have them exposed explicitly (i.e. wg_config, wg_offset, stopped_time)

File: sotodlib/wiregrid/calibrate.py
Comment (msilvafe) [inline sotodlib/wiregrid/calibrate.py:67]: I think call wg_config `ctx` or `wg_ctx`.

File: sotodlib/wiregrid/calibrate.py
Comment (msilvafe) [inline sotodlib/wiregrid/calibrate.py:91]: I think these 5 calls to get_detcosamp_hkaman should be doable in 1 if you pass a list to `fields` instead of a single field each time. Also the alias for lsl1, lsl2, lsr1, and lsr2 are all set to `limitswitchL1`

File: sotodlib/wiregrid/calibrate.py
Comment (msilvafe) [inline sotodlib/wiregrid/calibrate.py:46]: I think you probably want `in_place=True` here.

File: sotodlib/wiregrid/calibrate.py
Comment (msilvafe) [inline sotodlib/wiregrid/calibrate.py:49]: This function should probably have an `demodQ` and `demodU` as arguments. Typical behavior (as implemented in other modules) would be to have those arguments default set to None, then logic could be `if None` check if `demodQ` and `demodU` are in `aman` and if not either fail out or call the `hwp.demod` function.

File: sotodlib/wiregrid/calibrate.py
Comment (hnakata-JP) [reply sotodlib/wiregrid/calibrate.py:21]: Yeah, I have a question about that. What kind of style should we use when we call these wafers? I mean which notation should we use in the pipeline, ufm_mv19 or ws0?

File: sotodlib/wiregrid/calibrate.py
Comment (hnakata-JP) [reply sotodlib/wiregrid/calibrate.py:23]: Thanks. I will put all arguments explicitly.

File: sotodlib/wiregrid/calibrate.py
Comment (hnakata-JP) [reply sotodlib/wiregrid/calibrate.py:67]: Okay, i'll follow the customs.

File: sotodlib/wiregrid/calibrate.py
Comment (hnakata-JP) [reply sotodlib/wiregrid/calibrate.py:91]: I didn't know well this co-sample code. I'll check it to find a more simple description.

File: sotodlib/wiregrid/calibrate.py
Comment (msilvafe) [inline sotodlib/wiregrid/calibrate.py:214]: As described comment above, have demodQ/U passed as optional argument.

File: sotodlib/wiregrid/calibrate.py
Comment (hnakata-JP) [reply sotodlib/wiregrid/calibrate.py:46]: Ah, during the test phase, I wanted to keep the original one, but the same name was used lol. And it won't be needed any longer. Thanks.

File: sotodlib/wiregrid/calibrate.py
Comment (msilvafe) [inline sotodlib/wiregrid/calibrate.py:347]: You can use `aman.dets.count` instead of `np.shape(aman.det_info)[0]`

File: sotodlib/wiregrid/calibrate.py
Comment (msilvafe) [inline sotodlib/wiregrid/calibrate.py:246]: Explain in docstring what the len(x) == 2 and == 3 cases correspond to.

File: sotodlib/wiregrid/calibrate.py
Comment (hnakata-JP) [reply sotodlib/wiregrid/calibrate.py:49]: Yeah, this function requires that `demodQ` and `demodU` are included as attributes of aman. I have to implement a single `if` section on it.

File: sotodlib/wiregrid/calibrate.py
Comment (hnakata-JP) [reply sotodlib/wiregrid/calibrate.py:246]: Well, this is a little unclear section. I will give some explanations. Thank you.

File: sotodlib/wiregrid/calibrate.py
Comment (msilvafe) [inline sotodlib/wiregrid/calibrate.py:358]: You need to make a new label_axis in your axis manager with the name of the fit parameters (i.e. `A, B, C`) and when you wrap the cfitval, cfitcov, and cfitresvar you need to wrap them to those axes appropriately.

File: sotodlib/wiregrid/calibrate.py
Comment (msilvafe) [inline sotodlib/wiregrid/calibrate.py:393]: Same comment above about the label axis.

**Comment by hnakata-JP** [comment]
I have an overall question. Can I put the main code just below `sotodlib/` and make a code for updating the angle database under the `site_pipeline`?

File: sotodlib/wiregrid/calibrate.py
Comment (msilvafe) [reply sotodlib/wiregrid/calibrate.py:21]: Preference is for wafer_slot indexing as this is purely geometric and won't change between platforms or when we swap UFMs.

**Comment by msilvafe** [comment]
> I have an overall question. Can I put the main code just below `sotodlib/` and make a code for updating the angle database under the `site_pipeline`?

Possibly, I would have to defer to @mhasself or @kmharrington for that sort of organization. Also I'm not sure if I can give an approving review for this I think that probably has to come from one of these (or whichever DM lead oversees the WG code) which was why I provided only comments and not "review comments". 

Another point which I didn't know where exactly to put inline but it seems like there are a number of functions in here that loop over detectors (and steps). It would be worthwhile to take some time to think if you can matrix-ify any of those for loops as that tends to be the culprit for slowing a lot of this stuff down. I know that some of the fitting functions don't accept matrices so it may not be possible but worth looking into.

**Comment by kmharrington** [comment]
Before we go through and really review locations of code and things, can we get the configuration files and jupyter notebooks moved out? Right now I'm not completely sure what you already know you need to move versus what I should be telling you to change.

File: sotodlib/wiregrid/calibrate.py
Comment (kmharrington) [inline sotodlib/wiregrid/calibrate.py:28]: I don't think this docstring format ends up being parsed correctly by sphinx. Worth checking.

File: sotodlib/wiregrid/calibrate.py
Comment (kmharrington) [reply sotodlib/wiregrid/calibrate.py:90]: yes, it should be able to accept a slide of fields. And if it can't then we should fix it so it can.

File: sotodlib/wiregrid/calibrate.py
Comment (kmharrington) [inline sotodlib/wiregrid/calibrate.py:407]: This is un-necessary in here. (and remove the `#!/usr/bin/python3` from the top).  Things that should be run from command-line should go into `site-pipeline`.

File: sotodlib/wiregrid/calibrate.py
Comment (kmharrington) [inline sotodlib/wiregrid/calibrate.py:360]: don't need a pass at the end of a for loop

**Comment by kmharrington** [comment]
I think what's missing now is the script that will go into `site-pipeline` that will take an observation id of a wire grid test, run this analysis, and save its outputs?

File: sotodlib/wiregrid/calibrate.py
Comment (hnakata-JP) [reply sotodlib/wiregrid/calibrate.py:28]: Are there any sample scripts in sotodlib? I see some scripts have explanations, but others don't. I am not sure which one I should follow.

**Comment by hnakata-JP** [comment]
Sorry for the late update. I modified the structure of the files. Could somebody check to merge?

**Comment by msilvafe** [comment]
> Sorry for the late update. I modified the structure of the files. Could somebody check to merge?

This needs docs as well.

File: sotodlib/wiregrid/calibrate.py
Comment (msilvafe) [reply sotodlib/wiregrid/calibrate.py]: All of the docstrings in here (as one of many examples) are formatted correctly: https://github.com/simonsobs/sotodlib/blob/master/sotodlib/tod_ops/flags.py

I also typically test the docs build myself by running:
```$ virtualenv test
$ source test/bin/activate
$ pip install -r docs/requirements.txt
$ cd docs
$ make html```

That will check that the build works correctly (in addition to the automated github checks) and if it runs successfully will create a new directory called `_build` in the `docs` directory in sotodlib full of html files, you can open those in a browser to see how the formatting looks.

Comment (hnakata-JP) [reply sotodlib/wiregrid/calibrate.py]: Thanks. I'll try it soon.

**Comment by hnakata-JP** [comment]
I ask a stupid question about the documentation. I should modify `docs/site_pipeline.rst`, right?

**Review by mhasself — CHANGES_REQUESTED** [review]
Sorry it took me so long to look at this.  Couple of requests:
- Instead of `calibrate`, call the submodule `calibration`
- Put an empty `__init__.py` in the calibration/ directory.
- For docs, create docs/calibration.rst and initialize it as @msilvafe suggests
- Link calibration.rst into the main table of contents (`index.rst`), just after `site_pipeline.rst`

The docs are now being built automatically on push.  So you can check that for errors, and make sure the formatting looks good.  (Might need "Returns" instead of "Return", and might need to add more blank lines in some places... but not sure!)

This brings in `iminuit` as a new dependency.  That will break the docs build.  Add that to the list of mocked modules in docs/conf.py (search for `for missing in ...`).

**Comment by hnakata-JP** [comment]
When I am making documentation and compiling it, an issue occurs like the first attached image. I am not 100% sure if this issue is related to my documentation. Also, the documentation of the master branch is collapsed with a similar error. Can I avoid this? how should I do it?

スクリーンショット 2024-03-18 11 22 52
スクリーンショット 2024-03-18 11 26 16

**Comment by BrianJKoopman** [comment]
> When I am making documentation and compiling it, an issue occurs like the first attached image. I am not 100% sure if this issue is related to my documentation. Also, the documentation of the master branch is collapsed with a similar error. Can I avoid this? how should I do it?

Docs on `master` compile. The problem is in your branch. In `calibration.rst` you have `sotodlib.site_pipeline.calibration.wiregrid` when it should be `sotodlib.site_pipeline.calibrate.wiregrid`, the 'calibration' directory is 'calibrate'.

There are other problems with the build as well, but I'll let you dig into those.

**Comment by hnakata-JP** [comment]
Oh, is it caused by the lack of modification? I thought other modules made this. Thanks.

**Comment by hnakata-JP** [comment]
I frequently face a locale setting error when I run a command of `make html`. but this commitment should work for now. I will complete the documentation in another pull request, so can I merge this file first?

File: docs/calibration.rst
Comment (mhasself) [inline docs/calibration.rst:4]: remove this

File: docs/calibration.rst
Comment (mhasself) [inline docs/calibration.rst:11]: And remove these

File: sotodlib/site_pipeline/calibration/wiregrid.py
Comment (mhasself) [inline sotodlib/site_pipeline/calibration/wiregrid.py:54]: If user passed in telescope != None, then _telescope is not defined here.

File: sotodlib/site_pipeline/calibration/wiregrid.py
Comment (mhasself) [inline sotodlib/site_pipeline/calibration/wiregrid.py:60]: Make the hk_dir an optional argument for this function.

File: sotodlib/site_pipeline/calibration/wiregrid.py
Comment (mhasself) [inline sotodlib/site_pipeline/calibration/wiregrid.py:70]: Remove "pass" from non-empty blocks -- it does nothing.  I am seeing this in a lot of places...

File: sotodlib/site_pipeline/calibration/wiregrid.py
Comment (mhasself) [inline sotodlib/site_pipeline/calibration/wiregrid.py:116]: You are using `_telescope` and `telescope`; they are not consistent.

File: sotodlib/site_pipeline/calibration/wiregrid.py
Comment (mhasself) [inline sotodlib/site_pipeline/calibration/wiregrid.py:119]: It would probably be a good idea to store the _raw_ and _corrected_ values under different names.  E.g. `tod.wg.enc_rad = -tod.wg.enc_rad_raw + wg_offset`.  Will be less confusing, and that is "idempotent" (you can apply the correction multiple times and get the same result).

**Review by mhasself — CHANGES_REQUESTED** [review]
Thanks for getting this together.

A put a few comments about the code inline.

For the docs, note you can see a test build of this PR [here](https://sotodlib--699.org.readthedocs.build/en/699/calibration.html).   Go ahead and remove the "testing" stuff from calibration.rst and that's good enough I think.

**Comment by hnakata-JP** [comment]
Thank you for a lot of useful comments. I corrected them. I will merge this file in a day if there are other comments from other reviewers.

**Review by mhasself — APPROVED** [review]
Thanks!  (Let's give @msilvafe and @kmharrington a chance for last comments too.)

**Comment by kmharrington** [comment]
@guanyilun  will the presence of the submodule cause any issues for the so-workflow docker build? Would be good to make sure that will be ok. 

The files aren't at a point where we have argparsing / ability to integrate in with the automated pipeline but I'm happy to leave that for later work because I doubt we're ready for that anyway.

File: sotodlib/site_pipeline/calibration/wiregrid.py
Comment (msilvafe) [inline sotodlib/site_pipeline/calibration/wiregrid.py:196]: I think we don't want capital letters in function definitions.

File: sotodlib/site_pipeline/calibration/wiregrid.py
Comment (msilvafe) [inline sotodlib/site_pipeline/calibration/wiregrid.py:185]: enumerate?

File: sotodlib/site_pipeline/calibration/wiregrid.py
Comment (msilvafe) [inline sotodlib/site_pipeline/calibration/wiregrid.py:224]: `hasattr(tod, 'demodQ') is False` or not `hasattr(tod, 'demodQ')`

File: docs/index.rst
Comment (msilvafe) [inline docs/index.rst:26]: This is currently not being rendered properly. When I try to run I get this error:
```
WARNING: toctree contains reference to document 'calibration' that doesn't have a title: no link will be generated
```

I also it would be helpful if you added some documentation on what the module is doing, how to use it, how to run it in an automated/configuration file sense, etc. in a section of the calibration.rst called wiregrid. See some of the examples/notes/overviews in some of the other modules for examples.

**Comment by guanyilun** [comment]
> @guanyilun will the presence of the submodule cause any issues for the so-workflow docker build? Would be good to make sure that will be ok.

I don't see a main function in there, is there any reason we can't have the driver script outside of the submodule? That will be ideal. Though I don't think having it in submodule will break things, just have to name the module as `calibration.wiregrid`.

**Comment by hnakata-JP** [comment]
Thanks for the review, Max, Katie, and Yulin. I will integrate all your comments with another pull request because, at the beginning of this request, I was targeting the preparation of the wire grid functions for everyone's usage. Let's continue discussing this there.
Further, I will prepare the main function just under the `site_pipeline` to dump the calibration results as an HDF format to be loaded by `finalize_focal_plane.py`.

**Pull request merged**
